samples(Storage Batch Operations): Add samples and tests for storage batch operations

mahendra-google opened the pull request:
This PR contains samples and test cases to create, list, get, cancel and delete storage batch job operations.

The region tags used to create samples are as below:-

- [ ] 1. storage_batch_create_job 
- [ ] 2. storage_batch_list_jobs 
- [ ] 3. storage_batch_get_job
- [ ] 4. storage_batch_cancel_job
- [ ] 5. storage_batch_delete_job

mahendra-google committed:
samples(Storage Batch Operations): Add  samples and tests for storage batch operations

mahendra-google commented:
@purva9413  After looking at errors from build log , I think Storage Batch Operations API are not enabled on both Kokoro servers i.e. [Kokoro CI Linux 3](https://source.cloud.google.com/results/invocations/a7083ece-2df3-434b-9993-6e26f5b08e87) and [Kokoro CI Windows 3](https://source.cloud.google.com/results/invocations/efae90af-95c6-4280-8647-d9cc7aea4a70).

purva9413 commented:
The API has been enabled but there are failing tests. Please take a look and also please address comments from the review. Thanks!

mahendra-google commented on storagebatchoperations/api/runTests.ps1:
```diff
@@ -0,0 +1,16 @@
+// Copyright 2025 Google LLC
```
The commenting format has been revised to comply with coding standards. It is noted that the tests had already passed following earlier code changes involving the null-conditional operator (?.)."

mahendra-google commented:
@purva9413 Could you add once kokoro label? I want to check if the test is failing or not ?

purva9413 commented on storagebatchoperations/api/StorageBatchOperations.Samples.Tests/StorageFixture.cs:
```diff
@@ -0,0 +1,216 @@
+    /// Handles retry logic for operations that may fail due to transient errors,
+    /// such as network timeouts, service unavailability, or null reference exceptions.
+    /// </summary>
+    public class RetryRobot
```
You do not need to copy the entire class code in here. Please import the needed dependencies if any and only use the function for retrying. Similar to how it is done in the PR I linked earlier.

mahendra-google commented:
@purva9413 Could you add kokoro label again?.

mahendra-google committed:
fix(Storage Batch Operations): Modify retry logic

mahendra-google commented:
Closing this PR in favour of https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/3330

mahendra-google commented:
Reopned as new PR is not required

mahendra-google committed:
Update StorageBatchOperations.sln

mahendra-google committed:
Refactor CancelBatchJobTest for better error handling

Refactor CancelBatchJobTest to remove RetryRobot and handle exceptions.

mahendra-google committed:
Remove testutil project reference from tests

Removed ProjectReference to testutil.csproj.

mahendra-google committed:
Fix typo in README regarding .NET requirement

mahendra-google committed:
refactor(Storage Batch Operations): Remove unused variables and unused functions

mahendra-google committed:
Update CreateBatchJob.cs

mahendra-google committed:
refactor(Storage Batch Operations): Dispose memory stream

mahendra-google committed:
refactor(Storage Batch Operations): Convert create job test from fact to theory

mahendra-google committed:
fix(Storage Batch Operations): Add missing environment variables in fixture

mahendra-google committed:
refactor(Storage Batch Operations): Remove unused KMS fields in CreateBatchJobTest

Remove unused private fields `_kmsKey`, `_keyRingId`, `_cryptoKeyId`, and `_cryptoKeyName` from CreateBatchJobTest to improve code clarity and maintainability.

mahendra-google committed:
refactor(Storage Batch Operations): Update CancelBatchJobTest to improve job cancellation logic

mahendra-google committed:
refactor(Storage Batch Operations): Replace Thread.Sleep with loop in CancelBatchJobTest

mahendra-google committed:
style(Storage Batch Operations): Fix formatting in CancelBatchJobTest.cs

mahendra-google committed:
tests(Storage Batch Operations): Stabilize cancelBatchJobTest by increasing retry attempts

Extended the retry attempts in cancelBatchJobTest from 100 to 1000. This ensures the test accounts for delayed state transitions in the Koko environment while maintaining efficiency via an early exit once the 'Canceled' state is reached.

mahendra-google committed:
chore(Storage Batch Operations): Update dependencies


Updated Google.Cloud.Storage.V1 to 4.14.0 and Google.Cloud.StorageBatchOperations.V1 to 1.0.0-beta05.

krishnamd-jkp commented on storagebatchoperations/api/StorageBatchOperations.Samples/CreateBatchJob.cs:
```diff
@@ -0,0 +1,121 @@
+    /// <param name="jobTransformationObject">It is object of type either RewriteObject or PutMetadata.</param>
+    public Job CreateBatchJob(LocationName locationName,
+        BucketList bucketList,
+        string jobId = "12345678910",
```
Change this to <your-job-id>

mahendra-google replied:
**JobId** optional value is changed to 'your-job-id' in all **CreateBatchJob** samples

krishnamd-jkp commented on storagebatchoperations/api/StorageBatchOperations.Samples/CreateBatchJob.cs:
```diff
@@ -0,0 +1,121 @@
+        StorageBatchOperationsClient operationsClient = StorageBatchOperationsClient.Create();
+        Job job = null;
+        PutObjectHold putObjectHold;
+        switch (jobType)
```
It is better if we prefer small, focused snippets that do one thing well, rather than a single method that handles multiple operations based on string flags.

mahendra-google replied:
@krishnamd-jkp  To align with that approach, I have structured the latest updates so that **RewriteObject**, **DeleteObject**, **PutMetadata**, and **PutObjectHold** each have their own dedicated samples and corresponding tests. Instead of a single complex method, each operation is now isolated to ensure clarity and ease of use.

krishnamd-jkp commented on storagebatchoperations/api/StorageBatchOperations.Samples.Tests/CreateBatchJobTest.cs:
```diff
@@ -0,0 +1,121 @@
+    [InlineData("DeleteObject")]
+    [InlineData("PutObjectHold")]
+    [InlineData("RewriteObject")]
+    [InlineData("PutMetadata")]
```
If we split the CreateBatchJob sample into specific files for each operation, these tests can also be split into distinct, simpler test methods, which will be easier to maintain.

mahendra-google replied:
@krishnamd-jkp, I have split the tests for **RewriteObject**, **DeleteObject**, **PutMetadata**, and **PutObjectHold** into separate files to improve maintainability and clarity.

mahendra-google committed:
refactor(Storage Batch Operations): Addressing Review Feedback

mahendra-google committed:
refactor(Storage Batch Operations): Improve console output messages and update XML docs

iennae reviewed (approved):
approved by product engineer
